Infos über Requests sind hier:
https://2.python-requests.org/en/master/

In [1]:
import requests #wir laden die Bibliothek, und machen Python bereit. Am besten immer zuerst machen. 

In [70]:
import pandas as pd#jetzt laden wir die Bibliothek und vergeben ihm das kürzel pd

Wir schauen Erdbebendatenbank der USA an. Wir können dort gezielt Daten auslesen. 
https://earthquake.usgs.gov/fdsnws/event/1/#methods
Dort ist die API (Automatic Programmable Interface) dokumentiert. 
Man passt eigentlich die URL mit den entsprechenden Parametern an. 
Zum Beispiel die Methoden json und wadl. 
json ist ein Datenformat, ähnlich einem Python-Dictionary und eignet sich sehr gut für Python. 
wadl ist für Python etwas schwieriger zu verarbeiten. 

https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson
{"count":15125,"maxAllowed":20000}

Diese URL ist noch besser:
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2014-01-01&endtime=2014-01-02

URL können wir anpassen, zum Beispiel zur Suche in diesem September, bis heute. 
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-01&endtime=2019-09-23
10290 (Erdbeben registriert, seit Anfang September)

Twitter und Google haben sehr gute API's. Die von Facebook wird immer mehr geschlossen. 


In [4]:
r = requests.get("https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-01&endtime=2019-09-02")

Jetzt ist das Resultat der Abfrage in der Variablen r gespeichert.

In [5]:
r

<Response [200]>

In [8]:
type(r)

requests.models.Response

In [9]:
r.text

'644'

In [10]:
r.json()

644

In [11]:
Anzahl_Erdbeben = r.text

In [12]:
print(Anzahl_Erdbeben)

644


Jetzt bauen wir eine automatisierte Abfrage. 

In [22]:
url_anfang = "https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-0"
url_mitte = "&endtime=2019-09-0"

In [23]:
url_anfang + str(1) + url_mitte + str(2)

'https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-01&endtime=2019-09-02'

In [30]:
url_anfang = "https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-0"
url_mitte = "&endtime=2019-09-0"

for tag in range(1,5):
    print(url_anfang + str(tag) + url_mitte + str(tag + 1))

https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-01&endtime=2019-09-02
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-02&endtime=2019-09-03
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-03&endtime=2019-09-04
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-04&endtime=2019-09-05


Jetzt machen wir die Abfrage für requests und speichern die Resultate in die Liste (new_list):

In [31]:
url_anfang = "https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-0"
url_mitte = "&endtime=2019-09-0"
new_list = []
for tag in range(1,5):
    print(url_anfang + str(tag) + url_mitte + str(tag + 1))
    abfrage = requests.get(url_anfang + str(tag) + url_mitte + str(tag + 1))
    print(abfrage.text)
    new_list.append(abfrage.text)
    
    

https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-01&endtime=2019-09-02
644
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-02&endtime=2019-09-03
514
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-03&endtime=2019-09-04
607
https://earthquake.usgs.gov/fdsnws/event/1/count?starttime=2019-09-04&endtime=2019-09-05
512


In [32]:
print(new_list)

['644', '514', '607', '512']


In [35]:
for line in new_list:
    print(line)

644
514
607
512


jetzt könnte man das ganze noch direkt mit Datum ergänzen.  
print(now.strftime("%Y-%m-%d"))

Jetzt machen wir eine Abfrage nach Date (query):

In [36]:
abfrage = requests.get("https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2019-09-01&endtime=2019-09-02")

In [38]:
abfrage.text

'{"type":"FeatureCollection","metadata":{"generated":1569229628000,"url":"https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2019-09-01&endtime=2019-09-02","title":"USGS Earthquakes","status":200,"api":"1.8.1","count":644},"features":[{"type":"Feature","properties":{"mag":0.71999999999999997,"place":"28km E of Coso Junction, CA","time":1567382359230,"updated":1567382570608,"tz":-480,"url":"https://earthquake.usgs.gov/earthquakes/eventpage/ci39034144","detail":"https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=ci39034144&format=geojson","felt":null,"cdi":null,"mmi":null,"alert":null,"status":"automatic","tsunami":0,"sig":8,"net":"ci","code":"39034144","ids":",ci39034144,","sources":",ci,","types":",geoserve,nearby-cities,origin,phase-data,scitech-link,","nst":11,"dmin":0.109,"rms":0.22,"gap":103,"magType":"ml","type":"earthquake","title":"M 0.7 - 28km E of Coso Junction, CA"},"geometry":{"type":"Point","coordinates":[-117.64,36.064999999999998,1.29]},"id":"

Das Resultat ist ein langer String. Den können wir umwandeln:

In [39]:
abfrage = abfrage.json()

In [40]:
abfrage

{'type': 'FeatureCollection',
 'metadata': {'generated': 1569229628000,
  'url': 'https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2019-09-01&endtime=2019-09-02',
  'title': 'USGS Earthquakes',
  'status': 200,
  'api': '1.8.1',
  'count': 644},
 'features': [{'type': 'Feature',
   'properties': {'mag': 0.72,
    'place': '28km E of Coso Junction, CA',
    'time': 1567382359230,
    'updated': 1567382570608,
    'tz': -480,
    'url': 'https://earthquake.usgs.gov/earthquakes/eventpage/ci39034144',
    'detail': 'https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=ci39034144&format=geojson',
    'felt': None,
    'cdi': None,
    'mmi': None,
    'alert': None,
    'status': 'automatic',
    'tsunami': 0,
    'sig': 8,
    'net': 'ci',
    'code': '39034144',
    'ids': ',ci39034144,',
    'sources': ',ci,',
    'types': ',geoserve,nearby-cities,origin,phase-data,scitech-link,',
    'nst': 11,
    'dmin': 0.109,
    'rms': 0.22,
    'gap': 103,
    'magType

Welche Keys hat es überhaupt in diesen Datensätzen?
(Ergibt aber nur die oberste Ebene)

Und: wir haben ein Dictionary, das in Python so behandelt werden kann!

In [42]:
len(abfrage)

4

In [43]:
for keys in abfrage:
        print(keys)

type
metadata
features
bbox


In [44]:
abfrage["bbox"]

[-178.567, -49.9069, -3, 178.8008, 69.4703, 600.06]

In [45]:
abfrage["metadata"]

{'generated': 1569229628000,
 'url': 'https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2019-09-01&endtime=2019-09-02',
 'title': 'USGS Earthquakes',
 'status': 200,
 'api': '1.8.1',
 'count': 644}

Ok, das sind eigentlich Daten zu unserer Abfrage. Schauen wir uns den Key

In [46]:
abfrage["features"]

[{'type': 'Feature',
  'properties': {'mag': 0.72,
   'place': '28km E of Coso Junction, CA',
   'time': 1567382359230,
   'updated': 1567382570608,
   'tz': -480,
   'url': 'https://earthquake.usgs.gov/earthquakes/eventpage/ci39034144',
   'detail': 'https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=ci39034144&format=geojson',
   'felt': None,
   'cdi': None,
   'mmi': None,
   'alert': None,
   'status': 'automatic',
   'tsunami': 0,
   'sig': 8,
   'net': 'ci',
   'code': '39034144',
   'ids': ',ci39034144,',
   'sources': ',ci,',
   'types': ',geoserve,nearby-cities,origin,phase-data,scitech-link,',
   'nst': 11,
   'dmin': 0.109,
   'rms': 0.22,
   'gap': 103,
   'magType': 'ml',
   'type': 'earthquake',
   'title': 'M 0.7 - 28km E of Coso Junction, CA'},
  'geometry': {'type': 'Point', 'coordinates': [-117.64, 36.065, 1.29]},
  'id': 'ci39034144'},
 {'type': 'Feature',
  'properties': {'mag': 0.57,
   'place': '12km WSW of Searles Valley, CA',
   'time': 1567382032400,
   

In [48]:
erdbeben = abfrage["features"]

In [49]:
type(erdbeben)

list

Aha, jetzt haben wir eine Liste, eine Liste von Dictionaries!

In [52]:
erdbeben[0:4]

[{'type': 'Feature',
  'properties': {'mag': 0.72,
   'place': '28km E of Coso Junction, CA',
   'time': 1567382359230,
   'updated': 1567382570608,
   'tz': -480,
   'url': 'https://earthquake.usgs.gov/earthquakes/eventpage/ci39034144',
   'detail': 'https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=ci39034144&format=geojson',
   'felt': None,
   'cdi': None,
   'mmi': None,
   'alert': None,
   'status': 'automatic',
   'tsunami': 0,
   'sig': 8,
   'net': 'ci',
   'code': '39034144',
   'ids': ',ci39034144,',
   'sources': ',ci,',
   'types': ',geoserve,nearby-cities,origin,phase-data,scitech-link,',
   'nst': 11,
   'dmin': 0.109,
   'rms': 0.22,
   'gap': 103,
   'magType': 'ml',
   'type': 'earthquake',
   'title': 'M 0.7 - 28km E of Coso Junction, CA'},
  'geometry': {'type': 'Point', 'coordinates': [-117.64, 36.065, 1.29]},
  'id': 'ci39034144'},
 {'type': 'Feature',
  'properties': {'mag': 0.57,
   'place': '12km WSW of Searles Valley, CA',
   'time': 1567382032400,
   

In [53]:
len(erdbeben)

644

Also, wir haben 644 Erdbeben in dieser Liste verzeichnet. 

In [55]:
for dict_key in erdbeben[0]:
    print(dict_key)

type
properties
geometry
id


Jetzt gehen wir tiefer rein. Wir wollen jetzt Daten rausziehen, die uns interessieren. Zum Beispiel place, magnitude und time. Zuerst zum Testen mit print, um zu schauen, ob dabei rauskommt, was wir suchen. 

In [63]:
for beben in erdbeben:
    print(beben["properties"]["mag"])
    print(beben["properties"]["place"])
    print(beben["properties"]["time"])

0.72
28km E of Coso Junction, CA
1567382359230
0.57
12km WSW of Searles Valley, CA
1567382032400
1.6
30km WNW of Nikiski, Alaska
1567382016140
0.92
9km WNW of Calipatria, CA
1567382014290
0.77
14km E of Desert Hot Springs, CA
1567381965510
0.85
17km S of Searles Valley, CA
1567381893710
0.96
18km ESE of Little Lake, CA
1567381689840
1.78
19km S of Trona, CA
1567381242900
2.83
14km SSW of Searles Valley, CA
1567380906760
1.3
22km NNE of Badger, Alaska
1567380578357
1.5
6km NNW of Boron, CA
1567380269790
1.67
12km SSW of Searles Valley, CA
1567380135410
4.2
18km NE of Nikolski, Alaska
1567379957545
2.1
16km S of Fern Acres, Hawaii
1567379813380
0.34
97km NNW of Nikiski, Alaska
1567379479080
1.45
13km NNW of Mojave, CA
1567379408620
1.1
103km NW of Arctic Village, Alaska
1567379405574
1
82km W of Haines, Alaska
1567379001881
0.85
5km SW of Idyllwild, CA
1567378974490
1.45
19km ESE of Little Lake, CA
1567378876540
1.01
6km NNW of The Geysers, CA
1567378628100
2.1
7km WSW of Progreso, B.C.,

Jetzt bauen wir damit eine neue Liste:

In [66]:
eigeneliste = []
for beben in erdbeben:
    magnitude = beben["properties"]["mag"]
    place = beben["properties"]["place"]
    time = beben["properties"]["time"]
    mini_dict = {"Magnitude: ":magnitude, "Place: ":place, "Time: ":time}
    
    eigeneliste.append(mini_dict)

In [67]:
eigeneliste

[{'Magnitude: ': 0.72,
  'Place: ': '28km E of Coso Junction, CA',
  'Time: ': 1567382359230},
 {'Magnitude: ': 0.57,
  'Place: ': '12km WSW of Searles Valley, CA',
  'Time: ': 1567382032400},
 {'Magnitude: ': 1.6,
  'Place: ': '30km WNW of Nikiski, Alaska',
  'Time: ': 1567382016140},
 {'Magnitude: ': 0.92,
  'Place: ': '9km WNW of Calipatria, CA',
  'Time: ': 1567382014290},
 {'Magnitude: ': 0.77,
  'Place: ': '14km E of Desert Hot Springs, CA',
  'Time: ': 1567381965510},
 {'Magnitude: ': 0.85,
  'Place: ': '17km S of Searles Valley, CA',
  'Time: ': 1567381893710},
 {'Magnitude: ': 0.96,
  'Place: ': '18km ESE of Little Lake, CA',
  'Time: ': 1567381689840},
 {'Magnitude: ': 1.78,
  'Place: ': '19km S of Trona, CA',
  'Time: ': 1567381242900},
 {'Magnitude: ': 2.83,
  'Place: ': '14km SSW of Searles Valley, CA',
  'Time: ': 1567380906760},
 {'Magnitude: ': 1.3,
  'Place: ': '22km NNE of Badger, Alaska',
  'Time: ': 1567380578357},
 {'Magnitude: ': 1.5,
  'Place: ': '6km NNW of Boro

Und jetzt visualisieren wir diese Liste mit Pandas:

In [71]:
pd.DataFrame(eigeneliste)

Magnitude:                                  Place:          Time: 
0           0.72             28km E of Coso Junction, CA  1567382359230
1           0.57          12km WSW of Searles Valley, CA  1567382032400
2           1.60             30km WNW of Nikiski, Alaska  1567382016140
3           0.92               9km WNW of Calipatria, CA  1567382014290
4           0.77        14km E of Desert Hot Springs, CA  1567381965510
5           0.85            17km S of Searles Valley, CA  1567381893710
6           0.96             18km ESE of Little Lake, CA  1567381689840
7           1.78                     19km S of Trona, CA  1567381242900
8           2.83          14km SSW of Searles Valley, CA  1567380906760
9           1.30              22km NNE of Badger, Alaska  1567380578357
10          1.50                    6km NNW of Boron, CA  1567380269790
11          1.67          12km SSW of Searles Valley, CA  1567380135410
12          4.20             18km NE of Nikolski, Alaska  1567379957545
13          2.10            16km S of Fern Acres, Hawaii  1567379813380
14          0.34             97km NNW of Nikiski, Alaska  1567379479080
15          1.45                  13km NNW of Mojave, CA  1567379408620
16          1.10      103km NW of Arctic Village, Alaska  1567379405574
17          1.00                82km W of Haines, Alaska  1567379001881
18          0.85                 5km SW of Idyllwild, CA  1567378974490
19          1.45             19km ESE of Little Lake, CA  1567378876540
20          1.01              6km NNW of The Geysers, CA  1567378628100
21          2.10           7km WSW of Progreso, B.C., MX  1567378398760
22          1.12          14km SSW of Searles Valley, CA  1567378297040
23          0.26          95km SE of King Salmon, Alaska  1567377924510
24          4.90          43km SE of Sinabang, Indonesia  1567377683075
25          0.79            10km NE of Coso Junction, CA  1567377592020
26          0.70             45km NNW of Yakutat, Alaska  1567377084626
27          1.10                 71km S of Kobuk, Alaska  1567377073156
28          2.50                  12km NNW of Mojave, CA  1567377005020
29          0.55               6km NW of The Geysers, CA  1567377004500
..           ...                                     ...            ...
614         0.41                  1km S of Idyllwild, CA  1567300095980
615         0.80          98km SE of King Salmon, Alaska  1567299954032
616         1.61           7km WSW of Progreso, B.C., MX  1567299613810
617         1.00         6km WSW of North Nenana, Alaska  1567299567136
618         1.10      122km W of Haines Junction, Canada  1567299471569
619         0.60          25km N of North Nenana, Alaska  1567299388306
620         1.60     119km NNW of Arctic Village, Alaska  1567299268128
621         0.59               20km E of Little Lake, CA  1567299162980
622         1.10             116km SE of McGrath, Alaska  1567299054895
623         1.10    145km WNW of Haines Junction, Canada  1567298808409
624         1.18                  9km ESE of Olancha, CA  1567298562890
625         1.03            8km ENE of Coso Junction, CA  1567298178540
626         0.83            15km W of Searles Valley, CA  1567297840390
627        -0.40              45km ESE of Beatty, Nevada  1567297781864
628         0.85              15km ENE of Ridgecrest, CA  1567297766070
629        -0.50              46km ESE of Beatty, Nevada  1567297489885
630         1.32              12km ENE of Ridgecrest, CA  1567297344340
631         0.83             15km ESE of Little Lake, CA  1567297261580
632         1.78               6km SSE of Pahala, Hawaii  1567297170320
633         1.05                  9km ESE of Olancha, CA  1567297145560
634         1.83          11km SSE of Ponce, Puerto Rico  1567297114250
635         0.60      122km W of Haines Junction, Canada  1567297092005
636         0.80             19km ESE of Little Lake, CA  1567296940350
637         0.31                1km N of The 

Und jetzt speichern wir das Ergebnis noch ab, als .csv-Datei:

In [72]:
pd.DataFrame(eigeneliste).to_csv("erdbebenliste.csv")

Jetzt noch die Zeit umformatieren:
from datetime import datetime
datetime_object = datetime.strptime('Jun 1 2005  11:33', '%b %d %Y %H:%M')

oder damit:

from datetime import datetime.
timestamp = 1545730073.
dt_object = datetime. fromtimestamp(timestamp)
print("dt_object =", dt_object)
print("type(dt_object) =", type(dt_object))



In [78]:
import time #eine andere Bibliothek als datedime, etwas simpler.

eigeneliste = []

for beben in erdbeben:
    magnitude = beben["properties"]["mag"]
    place = beben["properties"]["place"]
    zeit = beben["properties"]["time"] #Achtung: time it jetzt auch die Funktion der Bibliothek time!
    zeit = time.ctime(int(zeit))
    mini_dict = {"Magnitude: ":magnitude, "Place: ":place, "Time: ":zeit}
    
    eigeneliste.append(mini_dict)

In [80]:
pd.DataFrame(eigeneliste)

Magnitude:                                  Place:   \
0           0.72             28km E of Coso Junction, CA   
1           0.57          12km WSW of Searles Valley, CA   
2           1.60             30km WNW of Nikiski, Alaska   
3           0.92               9km WNW of Calipatria, CA   
4           0.77        14km E of Desert Hot Springs, CA   
5           0.85            17km S of Searles Valley, CA   
6           0.96             18km ESE of Little Lake, CA   
7           1.78                     19km S of Trona, CA   
8           2.83          14km SSW of Searles Valley, CA   
9           1.30              22km NNE of Badger, Alaska   
10          1.50                    6km NNW of Boron, CA   
11          1.67          12km SSW of Searles Valley, CA   
12          4.20             18km NE of Nikolski, Alaska   
13          2.10            16km S of Fern Acres, Hawaii   
14          0.34             97km NNW of Nikiski, Alaska   
15          1.45                  13km NNW of Mojave, CA   
16          1.10      103km NW of Arctic Village, Alaska   
17          1.00                82km W of Haines, Alaska   
18          0.85                 5km SW of Idyllwild, CA   
19          1.45             19km ESE of Little Lake, CA   
20          1.01              6km NNW of The Geysers, CA   
21          2.10           7km WSW of Progreso, B.C., MX   
22          1.12          14km SSW of Searles Valley, CA   
23          0.26          95km SE of King Salmon, Alaska   
24          4.90          43km SE of Sinabang, Indonesia   
25          0.79            10km NE of Coso Junction, CA   
26          0.70             45km NNW of Yakutat, Alaska   
27          1.10                 71km S of Kobuk, Alaska   
28          2.50                  12km NNW of Mojave, CA   
29          0.55               6km NW of The Geysers, CA   
..           ...                                     ...   
614         0.41                  1km S of Idyllwild, CA   
615         0.80          98km SE of King Salmon, Alaska   
616         1.61           7km WSW of Progreso, B.C., MX   
617         1.00         6km WSW of North Nenana, Alaska   
618         1.10      122km W of Haines Junction, Canada   
619         0.60          25km N of North Nenana, Alaska   
620         1.60     119km NNW of Arctic Village, Alaska   
621         0.59               20km E of Little Lake, CA   
622         1.10             116km SE of McGrath, Alaska   
623         1.10    145km WNW of Haines Junction, Canada   
624         1.18                  9km ESE of Olancha, CA   
625         1.03            8km ENE of Coso Junction, CA   
626         0.83            15km W of Searles Valley, CA   
627        -0.40              45km ESE of Beatty, Nevada   
628         0.85              15km ENE of Ridgecrest, CA   
629        -0.50              46km ESE of Beatty, Nevada   
630         1.32              12km ENE of Ridgecrest, CA   
631         0.83             15km ESE of Little Lake, CA   
632         1.78               6km SSE of Pahala, Hawaii   
633         1.05                  9km ESE of Olancha, CA   
634         1.83          11km SSE of Ponce, Puerto Rico   
635         0.60      122km W of Haines Junction, Canada   
636         0.80             19km ESE of Little Lake, CA   
637         0.31                1km N of The Geysers, CA   
638         4.10  265km NNW of Raoul Island, New Zealand   
639         1.52                  23km NE of Oildale, CA   
640         2.03               11km NE of San Simeon, CA   
641         1.11                     11km S of Trona, CA   
642         4.10                  1km N of Norcia, Italy   
643         2.82        3km S of Leilani Estates, Hawaii   

                        Time:   
0    Sat May 15 14:40:30 51638  
1    Tue May 11 19:53:20 51638  
2    Tue May 11 15:22:20 51638  
3    Tue May 11 14:51:30 51638  
4    Tue May 11 01:18:30 51638  
5    Mon May 10 05:21:50 51638  
6    Fri May  7 20:44:00 51638  
7    Sun May  2 16:3